Sensor/icm42670 s aml

afontaine-invn opened the pull request:
Add AML algorithm 1.6.0 library support in driver and example
Through KConfig, configure different AML features:
- delta pointing, 
- detected gestures, 
- calibrated gyro offsets 
- or quaternion orientation

Commit:
Add AML algorithm Kconfig

Verfiy Init OK
libInvnAlgoAML_42670S.a generated by external_lib sample

Commit:
Print delta pointing output and detected gestures

Commit:
Define KConfig for different AML features

Report delta pointing or gestures (static, position, swipes)
AML library build in dedicated repo

Commit:
Also report AML output gyro offset and quaternion through KConfig

Commit:
Replace AML folder by the one generated from the 1.6.0 version

Commit:
Update README

rbuisson-invn commented on drivers/sensor/icm42670S/Kconfig:
```diff
@@ -33,6 +33,26 @@ config ICM42670S_APEX_WOM
+	bool "ICM42670S Air Motion Library (AML)"
+	default n
+
+choice AML_FEATURES
```
2 comments:
- Might be useful to explain that this controls what AML output is printed
- Could be interesting to allow multiple output at the same time

afontaine-invn replied:
Agreed, I will do the change. But I propose to keep by default only AML delta pointing, what do you think?

rbuisson-invn replied:
I think it's a wonderful idea, please proceed as you suggested.

rbuisson-invn commented on drivers/sensor/icm42670S/icm42670S.c:
```diff
@@ -224,6 +231,31 @@ static int icm42670S_channel_get(const struct device *dev,
+#ifdef CONFIG_ICM42670S_AML_POINTING
+	} else if ((enum sensor_channel_icm42670S)chan == SENSOR_CHAN_AML) {
+		val->val1 = data->delta[0];
+		val->val2 = data->delta[1];
```
This is not correct according to https://docs.zephyrproject.org/latest/hardware/peripherals/sensor.html#c.sensor_value
val2 is supposed to be the decimal part.

afontaine-invn replied:
Good catch I didn't know I wasn't allowed to use the structure differently. I will need to do the change also for APEX features outputs

rbuisson-invn commented on drivers/sensor/icm42670S/icm42670S.c:
```diff
@@ -224,6 +231,31 @@ static int icm42670S_channel_get(const struct device *dev,
+#ifdef CONFIG_ICM42670S_AML_GESTURES
+	} else if ((enum sensor_channel_icm42670S)chan == SENSOR_CHAN_AML) {
+		val->val1 = data->swipes_detected;
+		val->val2 = data->remote_position;
```
Same here, better to return several val than using val1 & val2

rbuisson-invn commented on drivers/sensor/icm42670S/icm42670S_aml.c:
```diff
@@ -0,0 +1,152 @@
+                                         -1,    0,    0,
+                                          0,    0,    1 };
+
+static int32_t gyro_mounting_matrix[9]= { 0,    1,    0,
```
Weird that gyro and accel mounting matric are not aligned, could you add a comment to explain?

afontaine-invn replied:
Good idea !

rbuisson-invn commented on drivers/sensor/icm42670S/icm42670S_aml.c:
```diff
@@ -0,0 +1,152 @@
+	struct icm42670S_data *data = dev->data;
+	
+	/* FSR */
+	config.acc_fsr = 16;
```
These values could be decided by the sample instead of hardcoded in the driver.
Not mandatory as setting max FSR for AML is a good practice.

rbuisson-invn commented on drivers/sensor/icm42670S/icm42670S_aml.c:
```diff
@@ -0,0 +1,152 @@
+	int16_t out_raw[3];
+	
+	for(i = 0; i < 3; i++) {
+		out_raw[i] =  ((int16_t)matrix[3*i+0] * raw[0]);
```
If you want to simplify the cast, you might just set the mounting matrix as int16_t I guess

rbuisson-invn commented on drivers/sensor/icm42670S/icm42670S_aml.c:
```diff
@@ -0,0 +1,152 @@
+	int64_t conv_val;
+
+	/* Converting q10 */
+	conv_val = (int64_t)(raw_val * 1000000) / (1 << 10);
```
This is loosing a bit of sensitivity I'm correct, but you comply with Zephyr sensor_value so :(...

rbuisson-invn commented on drivers/sensor/icm42670S/invn.algo.sw-integration.aml-42670S-gcc-zephyr-1.6.0/include/invn_algo_aml.h:
```diff
@@ -0,0 +1,169 @@
+}  RemotePosition;
+
+/* Forward declarations */
+struct inv_imu_device;
```
Something is missing here??
I would expect something like: struct inv_imu_device **imu**;

afontaine-invn replied:
It is a declaration for which the programmer has not yet given a complete definition. Indeed the structure is in the imu driver.

rbuisson-invn commented on samples/sensor/icm42670S/src/main.c:
```diff
@@ -120,6 +120,26 @@ int main(void)
 
 	printf("Configured for APEX data collecting.\n");
+#elif defined(CONFIG_ICM42670S_AML)
+	struct sensor_value aml_config;
```
Same comment as readout AML dX, dY should be an array of 2 val.

rbuisson-invn commented on samples/sensor/icm42670S/src/main.c:
```diff
@@ -214,6 +236,78 @@ int main(void)
+
+			if (aml_data[0].val1 != 0)
+				printf("[%s]: AML Gestures: %s \n",
+					   now_str(),			   
```
Gestures are not exclusive.
Should be 
if(aml_data[0].val1 && ICM42670S_AML_SWIPE_LEFT) printf("Swipe left");
if(aml_data[0].val1 && ICM42670S_AML_SWIPE_RIGHT) printf("Swipe right");
...

rbuisson-invn reviewed (approved):
No big issues, but please review and fix what you think is relevant before (or after) merging.

sriccardi-invn commented:
Could be good to add the AML doc

Commit:
Take PR remarks into account.

- review KConfig choice to have the possibility to enable all the features
- Properly use sensor_value structure
- Add documentation and comment about mounting matrix
- Fix gestures non exclusive
